### PR TITLE
[fleet_executor] Implementation of the message bus, the carrier and part of the interceptor

### PR DESCRIPTION
**Pull request by FeixLiu**
### PR types
Others

### PR changes
Others

### Describe
Implementation of the message bus, the carrier and part of the interceptor (communication relating part).

#### UT
Local test with `GLOG_v=3`
image

Note: this UT based on the ip of the network card, so I didn't push the test file to this PR. But do the test locally (as above). You can also test this on your machine [codes](https://gist.github.com/FeixLiu/449267f26341f477d3a33e9c6d7c149f). Replace this [CMakeLists](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/distributed/fleet_executor/CMakeLists.txt) with the CMakeLists in the Gist. Put the test.cc file in the Gist under `paddle/fluid/distributed/fleet_executor/` folder.

### CODE REVIEW

File: paddle/fluid/distributed/fleet_executor/carrier.cc [hunk 1]
```diff
@@ -21,22 +21,59 @@ namespace paddle {
...
+    : interceptor_id_to_node_(interceptor_id_to_node) {
+  CreateInterceptors();
 }
 
 Carrier::~Carrier() {
-  // destroy
+  for (auto iter = interceptor_idx_to_interceptor_.begin();
+       iter != interceptor_idx_to_interceptor_.end(); iter++) {
+    Interceptor* ip = iter->second.release();
+    delete ip;
```
Comment (LiYuRio) [inline paddle/fluid/distributed/fleet_executor/carrier.cc]: unique_ptr为什么要手动管理内存

File: paddle/fluid/distributed/fleet_executor/carrier.cc [hunk 2]
```diff
@@ -21,22 +21,59 @@ namespace paddle {
...
 }
 
 void Carrier::CreateInterceptors() {
   // create each Interceptor
+  for (auto iter = interceptor_id_to_node_.begin();
+       iter != interceptor_id_to_node_.end(); iter++) {
+    int64_t interceptor_id = iter->first;
+    TaskNode* task_node = iter->second;
+    interceptor_idx_to_interceptor_[interceptor_id] =
+        std::make_unique<Interceptor>(interceptor_id, task_node);
```
Comment (LiYuRio) [inline paddle/fluid/distributed/fleet_executor/carrier.cc]: 往map里插入用insert吧，下标访问包括了查找和插入两个语义，插入的时候判断一下id的唯一性

File: paddle/fluid/distributed/fleet_executor/interceptor.cc [hunk 3]
```diff
@@ -17,28 +17,73 @@
...
+      // local mailbox is empty, fetch the remote mailbox
+      VLOG(3) << interceptor_id_ << "'s local mailbox is empty. "
+              << "Fetch the remote mailbox.";
+      PADDLE_ENFORCE_EQ(FetchRemoteMailbox(), true,
+                        platform::errors::InvalidArgument(
+                            "Error encountered when fetch remote mailbox."));
+    }
+    const InterceptorMessage& interceptor_message = local_mailbox_.front();
+    local_mailbox_.pop();
+    const MessageType message_type = interceptor_message.message_type();
```
Comment (LiYuRio) [inline paddle/fluid/distributed/fleet_executor/interceptor.cc:64]: 返回const&？

File: paddle/fluid/distributed/fleet_executor/message_bus.cc [hunk 4]
```diff
@@ -12,42 +12,157 @@
...
+             "compiled with npu or Paddle isn't compiled "
+             "with distributed for now.";
+#endif
 }
 
 bool MessageBus::IsSameRank(int64_t src_id, int64_t dst_id) {
   // check whether the dst is the same rank or different rank with src
-  return true;
+  auto src_rank = interceptor_id_to_rank_.find(src_id);
+  auto dst_rank = interceptor_id_to_rank_.find(dst_id);
```
Comment (LiYuRio) [inline paddle/fluid/distributed/fleet_executor/message_bus.cc]: const auto&

File: paddle/fluid/distributed/fleet_executor/message_bus.cc [hunk 5]
```diff
@@ -12,42 +12,157 @@
...
+  PADDLE_ENFORCE_NE(
+      src_rank, interceptor_id_to_rank_.end(),
+      platform::errors::NotFound(
+          "Cannot find rank for src interceptor id %lld. Init error.", src_id));
+  PADDLE_ENFORCE_NE(
+      dst_rank, interceptor_id_to_rank_.end(),
+      platform::errors::NotFound(
+          "Cannot find rank for dst interceptor id %lld. Init error.", dst_id));
+  auto src_ip = rank_to_addr_.find(src_rank->second);
```
Comment (LiYuRio) [inline paddle/fluid/distributed/fleet_executor/message_bus.cc]: const auto&

File: paddle/fluid/distributed/fleet_executor/carrier.cc [hunk 6]
```diff
@@ -21,22 +21,59 @@ namespace paddle {
...
-  // init
+    const std::unordered_map<int64_t, TaskNode*>& interceptor_id_to_node)
+    : interceptor_id_to_node_(interceptor_id_to_node) {
+  CreateInterceptors();
 }
 
 Carrier::~Carrier() {
-  // destroy
+  for (auto iter = interceptor_idx_to_interceptor_.begin();
+       iter != interceptor_idx_to_interceptor_.end(); iter++) {
```
Comment (LiYuRio) [inline paddle/fluid/distributed/fleet_executor/carrier.cc]: 遍历可以简写，for(const auto& item : interceptor_idx_to_interceptor_)

File: paddle/fluid/distributed/fleet_executor/carrier.cc (hunk 1 above)
Comment (FeixLiu) [reply paddle/fluid/distributed/fleet_executor/carrier.cc]: 🤯🤯🤯

**Review by gongweibao — APPROVED** [review]
LGTM

**Pull request merged**